pkg/trace/api/status.go

```diff
@@ -0,0 +1,99 @@
+package api
+
+import (
+	"bytes"
+	"encoding/json"
+	"expvar"
+	"fmt"
+	"html/template"
+	"net/http"
+
+	"github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/dustin/go-humanize"
+)
+
+var statusTmpl = template.Must(template.New("status").Funcs(template.FuncMap{
+	"humanize": humanize.Commaf,
+	"percent":  func(v float64) string { return fmt.Sprintf("%02.1f", v*100) },
+}).Parse(`
+{{- if .error }}
+  Status: Not running or unreachable on localhost:{{.port}}.<br>
+  Error: {{.error}}<br>
+{{else}}
+  Status: Running<br>
+  Pid: {{.pid}}<br>
+  Uptime: {{.uptime}} seconds<br>
+  Mem alloc: {{humanize .memstats.Alloc}} bytes<br>
+  Hostname: {{.config.Hostname}}<br>
+  Receiver: {{.config.ReceiverHost}}:{{.config.ReceiverPort}}<br>
+  Endpoints:
+  <span class="stat_subdata">
+    {{- range $i, $e := .config.Endpoints}}
+    {{ $e.Host }}<br>
+    {{- end }}
+  </span>
+  <span class="stat_subtitle">Receiver (previous minute)</span>
+  <span class="stat_subdata">
+    {{- if .receiver -}}
+    {{- if eq (len .receiver) 0}}
+    No traces received in the previous minute.<br>
+    {{- end -}}
+    {{range $i, $ts := .receiver }}
+    From {{if $ts.Lang}}{{ $ts.Lang }} {{ $ts.LangVersion }} ({{ $ts.Interpreter }}), client {{ $ts.TracerVersion }}{{else}}unknown clients{{end}}<br>
+    <span class="stat_subdata">
+      Traces received: {{ $ts.TracesReceived }} ({{ humanize $ts.TracesBytes }} bytes)<br>
+      Spans received: {{ $ts.SpansReceived }}
+      {{ with $ts.WarnString }}
+      <br>WARNING: {{ . }}<br>
+      {{end}}
+    </span>
+    {{- end}}
+    {{- else -}}
+    Nothing received.<br />
+    {{- end -}}
+    {{range $key, $value := .ratebyservice -}}
+    {{- if eq $key "service:,env:" -}}
+    Default priority sampling rate: {{percent $value}}%
+    {{- else}}
+    Priority sampling rate for '{{ $key }}': {{percent $value}}%
+    {{- end}}
+    {{- end }}
+    {{- if lt .ratelimiter.TargetRate 1.0}}
+    <br>WARNING: Rate-limiter keep percentage: {{percent .ratelimiter.TargetRate}}%<br>
+    {{- end}}
+  </span>
+  <span class="stat_subtitle">Writer (previous minute)</span>
+  <span class="stat_subdata">
+    Traces: {{.trace_writer.Payloads}} payloads, {{.trace_writer.Traces}} traces, {{.trace_writer.Events}} events, {{.trace_writer.Bytes}} bytes<br>
+    {{- if gt .trace_writer.Errors 0.0}}WARNING: Traces API errors (1 min): {{.trace_writer.Errors}}{{end}}
+    Stats: {{.stats_writer.Payloads}} payloads, {{.stats_writer.StatsBuckets}} stats buckets, {{.stats_writer.Bytes}} bytes<br>
+    {{- if gt .stats_writer.Errors 0.0}}WARNING: Stats API errors (1 min): {{.stats_writer.Errors}}{{end}}
+  </span>
+{{end}}`))
+
+// reportStatus outputs the status in HTML for the web GUI.
+func (r *HTTPReceiver) reportStatus(w http.ResponseWriter, req *http.Request) {
```

Comment:
I'd have to look deeper into how we're serving the HTML on the browser side, but this made me think of a potential attack vector 🤔 

- kill the trace agent and/or intercept status request to the `apm_config.receiver_port` 
- serve code with malicious html+javascript to intercept auth token (for example)

I'm not sure if this would work to be honest, and you'd need to already have local access, but we'd have to be sure. Wondering if it might be better to respond with JSON.

Reply:
Yeah this was more like a test run to see if Maxime is happy with this approach. I've completely removed it and just compiled the template on the front-end.